model.py

```diff
@@ -1,103 +1,1222 @@
+        self.label = label
+
+        # print("Creating a new {}, labelled {}".format(
+        #     self.__class__.__name__, self.label))
+
+        # Check that the user defined their parameter names!
+        if len(self.node_par_names) != len(parameter_objects):
+            fail_msg = 'I recieved the wrong number of parameters!'
+            fail_msg += ' Expect: \n{}\nGot:\n{}'.format(
+                self.node_par_names,
+                [obj.name for obj in parameter_objects]
+            )
+            raise TypeError(fail_msg)
+
+        # Add the parameters to the self.XXX.
+        for par in parameter_objects:
+            setattr(self, par.name, par)
+
+        # Save the variable param names as a list
+        self.node_varpars = [par.name for par in parameter_objects
+                             if par.isVar]
+
+        # Handle the family
+        if children is None:
+            children = []
+        self.children = children
+        self.parent = parent
+
+        # Verify my parameters get put in the right places
+        self.__check_par_assignments__()
+
+    # Tree handling methods
+    def search_par(self, label, name):
+        '''Check if I have <name> in my attributes. If I don't, check my children.
+        If they don't, return None'''
+        # If I'm the desired node, get my parameter
+        if self.label == label:
+            return getattr(self, name)
+        # Otherwise, check my children.
+        else:
+            for child in self.children:
+                val = child.search_par(label, name)
+                if val is not None:
+                    return val
+            return None
+
+    def search_Node(self, class_type, label):
+        '''Search for a node below me of class_type, with the label requested.
+        '''
+        if self.name == "{}_{}".format(class_type, label):
+            return self
+        else:
+            for child in self.children:
+                val = child.search_Node(class_type, label)
+                if val is not None:
+                    return val
+                else:
+                    pass
+            return None
+
+    def add_child(self, children):
+        # This check allows XXX.add_child(Param) to be valid
+        if not isinstance(children, list):
+            children = [children]
+
+        new_children = self.children + children
+
+        self.children = new_children
+
+    # Tree evaluation methods
+    def chisq(self, plot=False):
+        '''Call the calc on each of my children. Overwrite this for the
+        bottom layer Model!'''
+
+        chisq = 0.0
+
+        if self.is_leaf:
+            if not hasattr(self, 'calc'):
+                print("{} has no calc function! Skipping...".format(self.name))
+            else:
+                calc = getattr(self, 'calc')
+                calc(plot)
+
+        for child in self.children:
+            try:
+                chisq += child.calc(plot)
```

Comment:
again, I'm a bit puzzled here, it looks like ```child.calc``` and ```child.chisq``` are interchangeable, so the logic has got a bit mixed up